src/python/pants/backend/python/target_types.py

```diff
@@ -393,6 +393,8 @@ class PexEnvField(DictStringToStringField):
 
 class PexPlatformsField(StringSequenceField):
     alias = "platforms"
+    removal_version = "2.20.0.dev0"
+    removal_hint = "Use complete_platforms instead."
```

Comment:
I think it might be nice to explain this a bit more, in two ways:

- link directly to the new multi-platform PEX docs (so a user trying to migrate is not having to first look at the `complete_platforms` doc string and then find the bigger-explanation docs from there) 
- either: expand those docs to discuss "why we're deprecating `platforms`" (temporarily, at least, e.g. something like https://www.pantsbuild.org/v2.18/docs/awslambda-python#migrating-from-pants-216-and-earlier) or have a sentence or two like that inline here.

Putting myself in the user's shoes, I would think that this migration will feel quite churn-ful for users that have have code that currently works ("this is working fine for me, why do I have to change"), and having easier links and understanding more background will, I hope, alleviate some of that frustration.

(Sorry for dribbling out these comments across multiple reviews!)

Reply:
I've shied away from this sort of thing and Pants as a whole, but completely agreed. Deprecations as they currently are used are a crutch in my mind that Pants should grow out of. They allow less than good decisions to continue to be made in general and mask the pain that should be felt by devs by shutting it to users.

Reply:
I agree that ~every deprecation represents a failure on our part to get an API decision right.  But the design space is huge, and as we feel our way through it we will make mistakes. 

To use this case as an example: we all agree that users shouldn't use `platforms`. That was a design mistake in Pex (or at least in post-Twitter Pex), that Pants carried over. There are two ways we can achieve this end:

A) Let user keep on trying to use `platforms`, occasionally failing in non-obvious ways, and coming to Slack for help, which takes up a lot of our time in providing the same advice over and over. 
B) A deprecation, which is flashing warning sign around the mistake that lets new users know proactively not to use it. 

We can debate how far out the deprecation should go. Certainly in this case it's easy to keep it for as long as we're on Pex 2 (my understanding is that Pex 3 won't support `platforms`?). That's one extreme. I lean towards taking away the bad thing within a reasonable time frame, with enough warning, reasonable people can disagree on that.

I do see your point about mitigating the frustration with better docs. I will look into that (after I fix another breakage caused by #19143)

Reply:
I'm not sure whether or not Pex 3 will support platforms. But I think you know I'm of the opinion taking away a thing that works for a user (@huon's point), is beyond maddening for that user. Doc, make non-default, etc - that's progress. Breaking an existing user is not. I'd not yank until Pants 3 - basically actually respect the side of semver that favors users looking for stability and not favoring developers of Pants.